--enable-fuzz fails in combination with --enable-wallet, --with-miniupnpc or --enable-zmq

`--enable-fuzz` fails in combination with `--enable-wallet`, `--with-miniupnpc` or `--enable-zmq`.

```
$ git clone https://github.com/bitcoin/bitcoin
$ cd bitcoin
$ ./autogen.sh
$ ./configure --disable-wallet --without-miniupnpc --disable-zmq --enable-fuzz
$ make
$ echo $?
0
$ make clean
$ ./configure --with-incompatible-bdb --without-miniupnpc --disable-zmq --enable-fuzz
$ make
…
  CXXLD    test/fuzz/address_deserialize
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `WalletInit::Construct(InitInterfaces&) const':
/…/src/wallet/init.cpp:139: undefined reference to `interfaces::MakeWalletClient(interfaces::Chain&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:4862: recipe for target 'test/fuzz/address_deserialize' failed
$ make clean
$ ./configure --disable-wallet --with-miniupnpc --disable-zmq --enable-fuzz
$ make
…
  CXXLD    test/fuzz/address_deserialize
libbitcoin_server.a(libbitcoin_server_a-net.o): In function `ThreadMapPort()':
/…/src/net.cpp:1412: undefined reference to `upnpDiscover'
/…/src/net.cpp:1423: undefined reference to `UPNP_GetValidIGD'
/…/src/net.cpp:1428: undefined reference to `UPNP_GetExternalIPAddress'
/…/src/net.cpp:1473: undefined reference to `freeUPNPDevlist'
/…/src/net.cpp:1475: undefined reference to `FreeUPNPUrls'
/…/src/net.cpp:1455: undefined reference to `UPNP_AddPortMapping'
/…/src/net.cpp:1461: undefined reference to `strupnperror'
/…/src/net.cpp:1467: undefined reference to `UPNP_DeletePortMapping'
/…/src/net.cpp:1469: undefined reference to `freeUPNPDevlist'
/…/src/net.cpp:1470: undefined reference to `FreeUPNPUrls'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
$ make clean
$ ./configure --disable-wallet --without-miniupnpc --enable-zmq --enable-fuzz
$ make
…
  CXXLD    test/fuzz/address_deserialize
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `Shutdown(InitInterfaces&)':
/…/src/init.cpp:290: undefined reference to `g_zmq_notification_interface'
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `SetupServerArgs()':
/…/src/init.cpp:469: undefined reference to `CZMQAbstractNotifier::DEFAULT_ZMQ_SNDHWM'
/…/src/init.cpp:470: undefined reference to `CZMQAbstractNotifier::DEFAULT_ZMQ_SNDHWM'
/…/src/init.cpp:471: undefined reference to `CZMQAbstractNotifier::DEFAULT_ZMQ_SNDHWM'
/…/src/init.cpp:472: undefined reference to `CZMQAbstractNotifier::DEFAULT_ZMQ_SNDHWM'
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `AppInitMain(InitInterfaces&)':
/…/src/init.cpp:1317: undefined reference to `RegisterZMQRPCCommands(CRPCTable&)'
/…/src/init.cpp:1440: undefined reference to `CZMQNotificationInterface::Create()'
/…/src/init.cpp:1440: undefined reference to `g_zmq_notification_interface'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

@fanquake @MarcoFalke I'm able to reproduce against current `master` (c7cfd20a77ce57d200b3b9e5e0dfb0d63818abdc) using a fresh clone:

```
$ git clone https://github.com/bitcoin/bitcoin bitcoin-enable-fuzz
$ cd bitcoin-enable-fuzz
$ git rev-parse HEAD
c7cfd20a77ce57d200b3b9e5e0dfb0d63818abdc
$ ./autogen.sh
$ ./configure --with-incompatible-bdb --without-miniupnpc --disable-zmq --enable-fuzz
$ make
…
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `WalletInit::Construct(InitInterfaces&) const':
/…/src/wallet/init.cpp:139: undefined reference to `interfaces::MakeWalletClient(interfaces::Chain&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >)'
collect2: error: ld returned 1 exit status
Makefile:4865: recipe for target 'test/fuzz/address_deserialize' failed
$ grep DESC /etc/lsb-release
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
$ gcc --version | head -1
gcc (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
$ ld --version | head -1
GNU ld (GNU Binutils for Ubuntu) 2.30
```

Just to rule out any local issue with my test system I reproduced the issue using a fresh 18.04 install too:

```
$ docker run -it ubuntu bash
$ apt update
$ apt upgrade
$ apt install gcc make git curl wget g++ build-essential libtool autotools-dev automake pkg-config bsdmainutils python3 libdb-dev libdb++-dev
$ apt install libssl-dev libevent-dev libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev
$ git clone https://github.com/bitcoin/bitcoin bitcoin-enable-fuzz
$ cd bitcoin-enable-fuzz
$ git rev-parse HEAD
c7cfd20a77ce57d200b3b9e5e0dfb0d63818abdc
$ ./autogen.sh
$ ./configure --with-incompatible-bdb --without-miniupnpc --disable-zmq --enable-fuzz
$ make
…
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `WalletInit::Construct(InitInterfaces&) const':
/bitcoin-enable-fuzz/src/wallet/init.cpp:139: undefined reference to `interfaces::MakeWalletClient(interfaces::Chain&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >)'
collect2: error: ld returned 1 exit status
Makefile:4865: recipe for target 'test/fuzz/address_deserialize' failed
$ grep DESC /etc/lsb-release
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
$ gcc --version | head -1
gcc (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
$ ld --version | head -1
GNU ld (GNU Binutils for Ubuntu) 2.30
```

Thanks for clarifying. I now see the same failure building inside `docker` with the steps above:
```bash
  CXX      bench/bench_bench_bitcoin-poly1305.o
  CXX      bench/bench_bench_bitcoin-prevector.o
  CXX      test/bench_bench_bitcoin-setup_common.o
  CXX      test/bench_bench_bitcoin-util.o
  CXX      bench/bench_bench_bitcoin-coin_selection.o
  CXX      bench/bench_bench_bitcoin-wallet_balance.o
  CXXLD    bench/bench_bitcoin
  CXX      test/test_fuzz_address_deserialize-setup_common.o
  CXX      test/fuzz/test_fuzz_address_deserialize-fuzz.o
  CXX      test/fuzz/test_fuzz_address_deserialize-deserialize.o
  CXXLD    test/fuzz/address_deserialize
libbitcoin_server.a(libbitcoin_server_a-init.o): In function `WalletInit::Construct(InitInterfaces&) const':
/bitcoin-enable-fuzz/src/wallet/init.cpp:139: undefined reference to `interfaces::MakeWalletClient(interfaces::Chain&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >)'
collect2: error: ld returned 1 exit status
Makefile:4865: recipe for target 'test/fuzz/address_deserialize' failed
make[2]: *** [test/fuzz/address_deserialize] Error 1
make[2]: Leaving directory '/bitcoin-enable-fuzz/src'
Makefile:12993: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/bitcoin-enable-fuzz/src'
Makefile:776: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
root@79e5030ee089:/bitcoin-enable-fuzz# gcc --version
gcc (Ubuntu 7.4.0-1ubuntu1~18.04) 7.4.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

root@79e5030ee089:/bitcoin-enable-fuzz# git rev-parse HEAD
c7cfd20a77ce57d200b3b9e5e0dfb0d63818abdc
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -165,7 +165,7 @@ jobs:
         RUN_FUNCTIONAL_TESTS=false
         RUN_FUZZ_TESTS=true
         GOAL="install"
-        BITCOIN_CONFIG="--disable-wallet --disable-bench --with-utils=no --with-daemon=no --with-libs=no --with-gui=no --enable-fuzz --with-sanitizers=fuzzer,address CC=clang CXX=clang++"
+        BITCOIN_CONFIG="--enable-fuzz --with-sanitizers=fuzzer,address CC=clang CXX=clang++"
 
     - stage: test
       name: 'x86_64 Linux  [GOAL: install]  [bionic]  [no wallet]'

--- a/configure.ac
+++ b/configure.ac
@@ -148,7 +148,8 @@ AC_ARG_ENABLE([extended-functional-tests],
     [use_extended_functional_tests=no])
 
 AC_ARG_ENABLE([fuzz],
-    AS_HELP_STRING([--enable-fuzz],[enable building of fuzz targets (default no)]),
+    AS_HELP_STRING([--enable-fuzz],
+    [enable building of fuzz targets (default no). enabling this will disable all other targets]),
     [enable_fuzz=$enableval],
     [enable_fuzz=no])
 
@@ -933,6 +934,29 @@ AC_SUBST(LEVELDB_CPPFLAGS)
 AC_SUBST(LIBLEVELDB)
 AC_SUBST(LIBMEMENV)
 
+dnl enable-fuzz should disable all other targets
+if test "x$enable_fuzz" = "xyes"; then
+  AC_MSG_WARN(enable-fuzz will disable all other targets)
+  build_bitcoin_utils=no
+  build_bitcoin_cli=no
+  build_bitcoin_tx=no
+  build_bitcoin_wallet=no
+  build_bitcoind=no
+  build_bitcoin_libs=no
+  bitcoin_enable_qt=no
+  bitcoin_enable_qt_test=no
+  bitcoin_enable_qt_dbus=no
+  enable_wallet=no
+  use_bench=no
+  use_upnp=no
+  use_zmq=no
+else
+  BITCOIN_QT_INIT
+
+  dnl sets $bitcoin_enable_qt, $bitcoin_enable_qt_test, $bitcoin_enable_qt_dbus
+  BITCOIN_QT_CONFIGURE([$use_pkgconfig])
+fi
+
 if test x$enable_wallet != xno; then
     dnl Check for libdb_cxx only if wallet enabled
     BITCOIN_FIND_BDB48
@@ -946,11 +970,6 @@ if test x$use_upnp != xno; then
     [have_miniupnpc=no]
   )
 fi
-
-BITCOIN_QT_INIT
-
-dnl sets $bitcoin_enable_qt, $bitcoin_enable_qt_test, $bitcoin_enable_qt_dbus
-BITCOIN_QT_CONFIGURE([$use_pkgconfig])
 
 if test x$build_bitcoin_wallet$build_bitcoin_cli$build_bitcoin_tx$build_bitcoind$bitcoin_enable_qt$use_tests$use_bench = xnonononononono; then
     use_boost=no

--- a/doc/fuzzing.md
+++ b/doc/fuzzing.md
@@ -46,7 +46,7 @@ export AFLPATH=$PWD
 To build Bitcoin Core using AFL instrumentation (this assumes that the
 `AFLPATH` was set as above):
 ```
-./configure --disable-ccache --disable-shared --enable-tests --enable-fuzz --disable-wallet --disable-bench --with-utils=no --with-daemon=no --with-libs=no --with-gui=no CC=${AFLPATH}/afl-gcc CXX=${AFLPATH}/afl-g++
+./configure --disable-ccache --disable-shared --enable-tests --enable-fuzz CC=${AFLPATH}/afl-gcc CXX=${AFLPATH}/afl-g++
 export AFL_HARDEN=1
 cd src/
 make
@@ -83,7 +83,7 @@ found in the `compiler-rt` runtime libraries package).
 To build all fuzz targets with libFuzzer, run
 
 ```
-./configure --disable-ccache --disable-wallet --disable-bench --with-utils=no --with-daemon=no --with-libs=no --with-gui=no --enable-fuzz --with-sanitizers=fuzzer,address CC=clang CXX=clang++
+./configure --disable-ccache --enable-fuzz --with-sanitizers=fuzzer,address CC=clang CXX=clang++
 make
 ```
 

